File: 993 - Cousins in Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isCousins(self, root: TreeNode, x: int, y: int) -> bool:
        self.parent = None
        depx, parentx = self.getDepth(root, x)
        depy, parenty = self.getDepth(root, y)
        return depx == depy and parentx != parenty
    
    def getDepth(self, root, val, curDep = 0, parent = None):
        if root is None:
            return -1, None
        if root.val == val:
            return curDep, parent
        d, p = self.getDepth(root.left, val, curDep + 1, root)
        if d == -1:
            d, p = self.getDepth(root.right, val, curDep + 1, root)
        return d, p

File: 993 - Cousins in Binary Tree.java
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode() {} TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) { this.val = val; this.left
 * = left; this.right = right; } }
 */
class Solution {
    private int depx = 0, depy = 0, xval, yval;
    private TreeNode parentx, parenty;
    
    public boolean isCousins(TreeNode root, int x, int y) {
        xval = x;
        yval = y;
        getDepth(root, 0, null);
        return depx == depy && parentx != parenty;
    }
    private void getDepth(TreeNode root, int curDep, TreeNode parent) {
        if (root == null) return;
        if (root.val == xval) {
            depx = curDep;
            parentx = parent;
        } else if (root.val == yval) {
            depy = curDep;
            parenty = parent;
        } else {
            getDepth(root.left, curDep + 1, root);
            getDepth(root.right, curDep + 1, root);
        }
    }
    
}